File: backend/exclusion/core/exclusion_reason.py
# SPDX-FileCopyrightText: 2026 Zampher
# SPDX-License-Identifier: MPL-2.0

"""
Exclusion reason enumeration for translation segments.
"""

from enum import Enum


class ExclusionReason(str, Enum):
    """
    Exclusion reason types for translation segments.
    
    Categories:
    - Content-based: Should always be excluded regardless of target language
    - Language-based: May change with target language
    - User-based: User manually selected
    """
    
    # Content-based exclusions (should always be excluded)
    IMAGE = "image"                    # Image placeholder or image content
    FORMULA = "formula"                # LaTeX/MathML formula
    REFERENCE = "reference"            # Reference/citation text
    IDENTIFIER = "identifier"          # URL, email, serial number, code, etc.
    STRUCTURAL = "structural"          # Header, footer, footnote, etc.
    
    # Optional exclusions (can be excluded by user choice, default: not excluded)
    TABLE = "table"                    # Table content (markdown table, PDF table block, etc.)
                                        # Default: not excluded (most tables can be translated)
                                        # User can choose to exclude tables if needed
    
    # Language-based exclusions (may change with target language)
    LANGUAGE_MATCH = "language_match"  # Source language matches target language
    
    # User-based exclusions
    USER_SELECTED = "user_selected"    # User manually selected to exclude
    
    # Legacy/unknown (for backward compatibility)
    UNKNOWN = "unknown"                # Unknown reason (legacy data)
    
    @classmethod
    def is_content_based(cls, reason: "ExclusionReason") -> bool:
        """Check if exclusion reason is content-based (should always be excluded)."""
        return reason in {
            cls.IMAGE,
            cls.FORMULA,
            cls.REFERENCE,
            cls.IDENTIFIER,
            cls.STRUCTURAL,
        }
        # Note: TABLE is NOT content-based - it's optional (user can choose to exclude)
    
    @classmethod
    def is_optional(cls, reason: "ExclusionReason") -> bool:
        """Check if exclusion reason is optional (can be excluded by user choice, default: not excluded)."""
        return reason == cls.TABLE

    @classmethod
    def is_default_not_excluded(cls, reason: "ExclusionReason") -> bool:
        """Reasons that are detected but not excluded by default; user chooses via checkbox (Structural, Language Match, Table)."""
        return reason in {cls.TABLE, cls.STRUCTURAL, cls.LANGUAGE_MATCH}

    @classmethod
    def is_language_based(cls, reason: "ExclusionReason") -> bool:
        """Check if exclusion reason is language-based (may change with target language)."""
        return reason == cls.LANGUAGE_MATCH
    
    @classmethod
    def is_user_based(cls, reason: "ExclusionReason") -> bool:
        """Check if exclusion reason is user-based."""
        return reason == cls.USER_SELECTED

    @classmethod
    def get_default_excluded(cls) -> set:
        """Return the set of ExclusionReasons that should be auto-excluded
        based on the ``exclusion_defaults`` section in system.json.

        Falls back to the original hard-coded defaults when the config
        cannot be loaded (e.g. during unit tests).
        """
        try:
            from config.system_config import get_system_config
            cfg = get_system_config().exclusion_defaults
            _mapping = {
                cls.IMAGE: cfg.image,
                cls.FORMULA: cfg.formula,
                cls.REFERENCE: cfg.reference,
                cls.IDENTIFIER: cfg.identifier,
                cls.STRUCTURAL: cfg.structural,
                cls.TABLE: cfg.table,
                cls.LANGUAGE_MATCH: cfg.language_match,
            }
            return {reason for reason, enabled in _mapping.items() if enabled}
        except Exception:
            # Fallback: original hard-coded defaults
            return {cls.IMAGE, cls.FORMULA, cls.REFERENCE, cls.IDENTIFIER}

File: frontend/lib/features/translation/models/exclusion_reason.dart
// SPDX-FileCopyrightText: 2025 QinHan
// SPDX-License-Identifier: MPL-2.0

import 'package:flutter/material.dart';
import '../../../l10n/app_localizations.dart';

/// Exclusion reason enum for translation segments
/// Maps backend exclusion_reason values to frontend display properties
enum ExclusionReason {
  image('image', 'Image', Icons.image, Colors.blue),
  formula('formula', 'Formula', Icons.functions, Colors.purple),
  table('table', 'Table', Icons.table_chart, Colors.teal),
  reference('reference', 'Reference', Icons.format_quote, Colors.orange),
  identifier('identifier', 'Identifier', Icons.tag, Colors.grey),
  structural('structural', 'Structural', Icons.view_headline, Colors.brown),
  languageMatch(
      'language_match', 'Language Match', Icons.language, Colors.green,),
  userSelected('user_selected', 'User Excluded', Icons.block, Colors.red),
  unknown('unknown', 'Excluded', Icons.block, Colors.grey);

  final String value;
  final String displayName;
  final IconData icon;
  final Color color;

  const ExclusionReason(
    this.value,
    this.displayName,
    this.icon,
    this.color,
  );

  /// Localized display name for UI (use instead of [displayName] when l10n is available).
  String displayNameLocalized(AppLocalizations l10n) {
    switch (this) {
      case ExclusionReason.image:
        return l10n.settingsExclusionImageTitle;
      case ExclusionReason.formula:
        return l10n.settingsExclusionFormulaTitle;
      case ExclusionReason.table:
        return l10n.settingsExclusionTableTitle;
      case ExclusionReason.reference:
        return l10n.settingsExclusionReferenceTitle;
      case ExclusionReason.identifier:
        return l10n.settingsExclusionIdentifierTitle;
      case ExclusionReason.structural:
        return l10n.settingsExclusionStructuralTitle;
      case ExclusionReason.languageMatch:
        return l10n.settingsExclusionLanguageMatchTitle;
      case ExclusionReason.userSelected:
        return l10n.exclusionPanelUserExcluded;
      case ExclusionReason.unknown:
        return l10n.exclusionPanelExcluded;
    }
  }

  /// Convert string value to ExclusionReason enum
  /// Returns unknown if value is null or not found
  static ExclusionReason fromString(String? value) {
    if (value == null) return unknown;
    return ExclusionReason.values.firstWhere(
      (e) => e.value == value,
      orElse: () => unknown,
    );
  }

  /// Check if this exclusion reason is content-based (cannot be unexcluded)
  /// Content-based exclusions: image, formula, identifier, reference, structural
  /// Note: table is NOT content-based - it's optional (user can choose to exclude)
  bool get isContentBased =>
      this == image ||
      this == formula ||
      this == identifier ||
      this == reference ||
      this == structural;

  /// Check if this exclusion reason is language-based
  bool get isLanguageBased => this == languageMatch;

  /// Check if this exclusion reason is user-based (can be unexcluded)
  bool get isUserBased => this == userSelected;

  /// Check if this exclusion can be unexcluded by user
  /// user_selected, language_match, identifier, table, and formula can be unexcluded
  /// formula, identifier and table can be modified by user (e.g., if they were incorrectly detected)
  bool get canUnexclude =>
      isUserBased ||
      isLanguageBased ||
      this == identifier ||
      this == table ||
      this == formula;
}
